zebra-rpc/src/methods/get_block_template_rpcs/zip317.rs

```diff
@@ -201,20 +201,13 @@ fn checked_add_transaction_weighted_random(
 /// If some transactions have not yet been chosen, returns the weighted index and the transaction.
 /// Otherwise, just returns the transaction.
 fn choose_transaction_weighted_random(
-    candidate_txs: &[VerifiedUnminedTx],
-    mut weighted_index: WeightedIndex<f32>,
+    candidate_txs: &mut Vec<VerifiedUnminedTx>,
+    weighted_index: WeightedIndex<f32>,
 ) -> (Option<WeightedIndex<f32>>, VerifiedUnminedTx) {
     let candidate_position = weighted_index.sample(&mut thread_rng());
-    let candidate_tx = candidate_txs[candidate_position].clone();
+    let candidate_tx = candidate_txs.swap_remove(candidate_position);
 
     // Only pick each transaction once, by setting picked transaction weights to zero
-    if weighted_index
-        .update_weights(&[(candidate_position, &0.0)])
-        .is_err()
-    {
-        // All weights are zero, so each transaction has either been selected or rejected
-        (None, candidate_tx)
-    } else {
-        (Some(weighted_index), candidate_tx)
-    }
+    // All weights are zero, so each transaction has either been selected or rejected
+    (setup_fee_weighted_index(&candidate_txs), candidate_tx)
```

Comment:
Let's explain why we have to regenerate the index every time, so we don't try to optimise this code in future, and re-introduce the bug.

```suggestion
    // We have to regenerate this index each time we choose a transaction, due to floating-point sum inaccuracies.
    // If we don't, some chosen transactions can end up with a tiny non-zero weight, leading to duplicates.
    // <https://github.com/rust-random/rand/blob/4bde8a0adb517ec956fcec91665922f6360f974b/src/distributions/weighted_index.rs#L173-L183>
    (setup_fee_weighted_index(&candidate_txs), candidate_tx)
```

Reply:
Good idea, thank you. Applied in [`b1cec4f`](https://github.com/ZcashFoundation/zebra/pull/6006/commits/b1cec4fb44d64521b16adca1532ead59a9104d51).

Reply:
Btw I've never liked that zcashd uses floating point here, and that is something we might change. Let us know if you want to change it to integer or fixed point arithmetic.

Reply:
We could use fixed point with u64, and scale the fee by 2^13 before dividing. That would get us close to the maximum number of actions. (Which I think is roughly 300K for transparent outputs.)

But we don't need to make that change any time soon, we're focused on finishing off all the RPCs that miners need.